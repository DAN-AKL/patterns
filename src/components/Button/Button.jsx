import React from "react";
import * as PropTypes from "prop-types";

import "./Button.scss";

const propTypes = {
  as: PropTypes.oneOf(["button", "anchor"]),
  children: PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  href: PropTypes.string,
  icon: PropTypes.object,
  icon_placement: PropTypes.oneOf(["before", "after"]),
  onChange: PropTypes.func,
  priority: PropTypes.oneOf(["primary", "secondary", "tertiary"]),
  small: PropTypes.bool
};

const defaultProps = {
  as: "button",
  disabled: false,
  icon_placement: "after",
  priority: "primary"
};

class Button extends React.Component {
  constructor(props) {
    super(props);
  }

  common_props() {
    return {
      onClick: this.props.onClick,
      disabled: this.props.disabled,
      className: `
        btn
        btn--${this.props.priority}
        ${this.props.className ? this.props.className : ""}
        ${this.props.small ? "btn--small" : ""}
      `
    };
  }

  common_markup() {
    return (
      <>
        {this.props.icon_placement === "before" && this.props.icon}
        <span className="btn__text">{this.props.children}</span>
        {this.props.icon_placement === "after" && this.props.icon}
      </>
    );
  }

  button_markup() {
    return <button {...this.common_props()}>{this.common_markup()}</button>;
  }

  anchor_markup() {
    return (
      <a href={this.props.href} {...this.common_props()}>
        {this.common_markup()}
      </a>
    );
  }

  render() {
    switch (this.props.as) {
      case "anchor":
        return this.anchor_markup();
        break;

      case "button":
        return this.button_markup();
        break;

      default:
        return this.button_markup();
    }
  }
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
