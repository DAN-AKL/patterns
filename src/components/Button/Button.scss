.btn {
  align-items: center;
  border: 2px solid currentColor;
  color: currentColor;
  cursor: pointer;
  display: inline-flex;
  font-size: 5rem;
  font-weight: 800;
  justify-content: left;
  line-height: 1;
  padding: 3rem 5rem;
  text-align: left;
  text-decoration: none;
  transition: background-color 250ms ease-out;

  svg {
    fill: currentColor;
    height: 6rem;
    margin-left: 4rem;
    margin-top: 0;
    width: 6rem;
  }
}

.btn__text {
  flex-shrink: 1;
  flex-grow: 0;
}

.btn--small {
  font-size: 4rem;
  font-weight: 400;
  padding: 2rem 3rem;
}

.btn--primary {
  background-color: hsl(46, 98%, 59%);
  color: #000;

  &:hover {
    background-color: hsl(46, 100%, 48%);
  }

  &[disabled] {
    cursor: not-allowed;
    background-color: hsl(46, 98%, 86%);
    border-color: hsl(0, 20%, 74.5%);
    color: hsl(0, 20%, 74.5%);

    &:hover {
      background-color: hsl(46, 98%, 86%);
      border-color: hsl(0, 20%, 74.5%);
      color: hsl(0, 20%, 74.5%);
    }
  }
}

.btn--secondary {
  background-color: hsl(46, 98%, 86%);

  &[disabled] {
    cursor: not-allowed;
    background-color: hsl(46, 98%, 86%);
    border-color: hsl(0, 20%, 74.5%);
    color: hsl(0, 20%, 74.5%);

    &:hover {
      background-color: hsl(46, 98%, 86%);
      border-color: hsl(0, 20%, 74.5%);
      color: hsl(0, 20%, 74.5%);
    }
  }
}

.btn--tertiary {
  background-color: transparent;
  border-color: transparent;
  padding: 0;
}
