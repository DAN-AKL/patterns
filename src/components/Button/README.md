# Header

dkfjghldfskjghdflkjdhf

- dfkgjhdfg
- sdkgjhdfgk
- sdfgjkldfkgj
