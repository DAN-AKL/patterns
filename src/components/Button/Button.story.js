import React from "react";

import Button from "./Button.jsx";
import notes from "./README.md";

import Arrow from "../../images/icons/arrows/arrow-right.svg";
import Chevron from "../../images/icons/arrows/chevron-right.svg";
import Menu from "../../images/icons/basic/menu.svg";

import { storiesOf } from "@storybook/react";

import {
  withKnobs,
  text,
  boolean,
  number,
  select
} from "@storybook/addon-knobs";

import { action } from "@storybook/addon-actions";
// import figmaDecorator from "storybook-addon-figma";

const stories = storiesOf("Components/Button", module);

stories.addDecorator(withKnobs);

const priorities = {
  Primary: "primary",
  Secondary: "secondary",
  Tertiary: "tertiary"
};

const as = {
  Button: "button",
  Anchor: "anchor"
};

const icons = {
  None: "",
  Arrow: "Arrow",
  Chevron: "Chevron",
  Menu: "Menu"
};

const icon_map = {
  None: "",
  Arrow: <Arrow />,
  Chevron: <Chevron />,
  Menu: <Menu />
};

const icon_placements = {
  Before: "before",
  After: "after"
};

stories.add(
  "Default",
  () => {
    return (
      <Button
        priority={select("Priority", priorities, "primary")}
        as={select("As", as, "button")}
        small={boolean("Small", false)}
        disabled={boolean("Disabled", false)}
        icon={icon_map[select("Icon", icons, "")]}
        icon_placement={select("Icon placement", icon_placements, "after")}
        onClick={action("Button click")}
      >
        {text("Button text", "I am a Button")}
      </Button>
    );
  },
  { notes: notes }
);
