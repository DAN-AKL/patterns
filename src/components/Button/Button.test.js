import React from "react";
import renderer from "react-test-renderer";
import Button from "./Button.jsx";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Button priority="primary" as="button">
        I am a button
      </Button>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
