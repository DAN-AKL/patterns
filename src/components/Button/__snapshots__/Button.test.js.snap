// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="
        btn
        btn--primary
        
        
      "
>
  <span
    className="btn__text"
  >
    I am a button
  </span>
</button>
`;
